pkg/controller/util/endpoint/deduplicating_rate_limiting_queue_test.go

```diff
@@ -0,0 +1,156 @@
+	"k8s.io/client-go/util/workqueue"
+)
+
+func waitForQueueNonEmpty(queue *DeduplicatingRateLimitingQueue) error {
+	return wait.PollImmediate(time.Millisecond, 100*time.Millisecond, func() (bool, error) {
+		return queue.Len() > 0, nil
+	})
+}
+
+func expectKey(t *testing.T, queue *DeduplicatingRateLimitingQueue, want string) {
+	if err := waitForQueueNonEmpty(queue); err != nil {
+		t.Fatalf("expected at least one element: %v", err)
+	}
+
+	if got, _ := queue.Get(); got != want {
+		t.Fatalf("queue.Get() = %v; want: %v", got, want)
+	}
+	queue.Done(want)
+	queue.Forget(want)
+}
+
+func TestBasic(t *testing.T) {
+	clock := clock.NewFakeClock(time.Now())
+	queue := NewDeduplicatingRateLimitingQueue(
+		workqueue.NewNamedRateLimitingQueueWithCustomClock(workqueue.DefaultControllerRateLimiter(), "", clock),
+		// There is no way to mock clock in NewItemBucketRateLimiter,
+		// so we are not able to test a case when we are stopped being rate limited.
+		workqueue.NewItemBucketRateLimiter(rate.Limit(1.0), 1),
+	)
+	defer queue.ShutDown()
+
+	// First call without delay.
+	queue.Enqueue("key1", true)
+
+	expectKey(t, queue, "key1")
+
+	// key1 should be rate limited now
+	queue.Enqueue("key1", true)
+	queue.Enqueue("key2", true)
+
+	// key1 is added with 1s delay, so key2 is only available key now.
+	expectKey(t, queue, "key2")
+
+	clock.Step(1 * time.Second)
+	expectKey(t, queue, "key1")
+}
+
+func TestDedup(t *testing.T) {
+	clock := clock.NewFakeClock(time.Now())
+	queue := NewDeduplicatingRateLimitingQueue(
+		workqueue.NewNamedRateLimitingQueueWithCustomClock(workqueue.DefaultControllerRateLimiter(), "", clock),
+		// There is no way to mock clock in NewItemBucketRateLimiter,
+		// so we are not able to test a case when we are stopped being rate limited.
+		workqueue.NewItemBucketRateLimiter(rate.Limit(1.0), 1),
+	)
+	defer queue.ShutDown()
+
+	// We add key1 three times. The expected result is that it will be deduped.
+	queue.Enqueue("key1", false)
+	queue.Enqueue("key1", false)
+	queue.Enqueue("key1", false)
+
+	queue.Enqueue("key2", false)
+
+	expectKey(t, queue, "key1")
+	expectKey(t, queue, "key2")
+}
+
+func TestAllowRateLimited(t *testing.T) {
+	clock := clock.NewFakeClock(time.Now())
+	queue := NewDeduplicatingRateLimitingQueue(
+		workqueue.NewNamedRateLimitingQueueWithCustomClock(workqueue.DefaultControllerRateLimiter(), "", clock),
+		// There is no way to mock clock in NewItemBucketRateLimiter,
+		// so we are not able to test a case when we are stopped being rate limited.
+		workqueue.NewItemBucketRateLimiter(rate.Limit(1.0), 1),
+	)
+	defer queue.ShutDown()
+
+	// First call without delay.
+	queue.Enqueue("key1", true)
+
+	expectKey(t, queue, "key1")
+
+	// key1 should be rate limited now
+	queue.Enqueue("key1", false)
+	queue.Enqueue("key2", true)
+
+	// key1 has been added with allowRateLimit false, so it's not rate limited.
+	expectKey(t, queue, "key1")
+	expectKey(t, queue, "key2")
+}
+
+func TestEnqueueRetry(t *testing.T) {
```

Comment:
Not sure if this is worth testing:

A combination of `EnqueueRetry(key)` and `Enqueue(key, true)`, `Enqueue(key, false)`. 

Ideally, would love to have a clear defined behavior.

Reply:
I tried to read workqueue.RateLimitingInterface code and if I understand that correctly, you can have up to two elements queued:
* one added by Add
* one added by AddAfter

Moreover,
a series of AddAfter(1 * time.Second) and AddAfter(500 * time.Millisecond) calls will result in one element queued after 500ms while
AddAfter(1 * time.Second) and AddAfter(0) will result in two elements: one after 1s and one now (which is quite surprising for me).

@lavalamp , could you confirm?

If this is true, then my attempt to trace elements in queue (the notStarted map) is incorrect: it will be cleared on the first element run, while the second can be still in queue.